rpcperms/interceptor.go

```diff
@@ -593,3 +707,202 @@ func (r *InterceptorChain) rpcStateStreamServerInterceptor() grpc.StreamServerIn
 		return handler(srv, ss)
 	}
 }
+
+// middlewareUnaryServerInterceptor is an unary gRPC interceptor that intercepts
+// all requests and responses that are send with a macaroon containing a custom
+// caveat condition that is handled by registered middleware.
+func (r *InterceptorChain) middlewareUnaryServerInterceptor() grpc.UnaryServerInterceptor {
+	return func(ctx context.Context,
+		req interface{}, info *grpc.UnaryServerInfo,
+		handler grpc.UnaryHandler) (interface{}, error) {
+
+		msg, err := NewMessageInterceptionRequest(
+			ctx, TypeRequest, false, info.FullMethod, req,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		err = r.acceptRequest(msg)
+		if err != nil {
+			return nil, err
+		}
+
+		resp, respErr := handler(ctx, req)
+		if respErr != nil {
+			return resp, respErr
+		}
+
+		return r.interceptResponse(ctx, false, info.FullMethod, resp)
+	}
+}
+
+// middlewareStreamServerInterceptor is a streaming gRPC interceptor that
+// intercepts all requests and responses that are send with a macaroon
+// containing a custom caveat condition that is handled by registered
+// middleware.
+func (r *InterceptorChain) middlewareStreamServerInterceptor() grpc.StreamServerInterceptor {
+	return func(srv interface{},
+		ss grpc.ServerStream, info *grpc.StreamServerInfo,
+		handler grpc.StreamHandler) error {
+
+		// Don't intercept the interceptor itself which is a streaming
+		// RPC too!
+		if info.FullMethod == lnrpc.RegisterRPCMiddlewareURI {
+			return handler(srv, ss)
+		}
+
+		// To give the middleware a chance to accept or reject the
+		// establishment of the stream itself (and not only when the
+		// first message is sent on the stream), we send an intercept
+		// request for the stream auth now:
+		msg, err := NewStreamAuthInterceptionRequest(
+			ss.Context(), info.FullMethod,
+		)
+		if err != nil {
+			return err
+		}
+
+		err = r.acceptRequest(msg)
+		if err != nil {
+			return err
+		}
+
+		wrappedSS := &serverStreamWrapper{
+			ServerStream: ss,
+			fullMethod:   info.FullMethod,
+			interceptor:  r,
+		}
+
+		return handler(srv, wrappedSS)
+	}
+}
+
+// acceptRequest sends an intercept request to all middlewares that have
+// registered for it. This means either a middleware has requested read-only
+// access or the request actually has a macaroon which a caveat the middleware
+// registered for.
+func (r *InterceptorChain) acceptRequest(msg *InterceptionRequest) error {
+	r.RLock()
+	defer r.RUnlock()
+
+	for _, middleware := range r.registeredMiddleware {
```

Comment:
Hmm, don't we only want to send to the middleware the precisely matches the specified custom caveat?

Reply:
We give each request to every middleware that qualifies. There can only be one middleware per unique custom caveat name. But there is also the concept of "read-only" middleware that can inspect _ever_ request (but can never manipulate a response). So a request can be sent out to multiple middlewares sequentially. The `if` below checks what middleware qualifies.